b2sdk/transfer/inbound/downloaded_file.py

```diff
@@ -0,0 +1,149 @@
+        self.mod_time_to_set = mod_time
+
+    def write(self, value):
+        """
+        This method is overwritten (monkey-patched) in __enter__ for performance reasons
+        """
+        raise NotImplementedError
+
+    def read(self, *a):
+        """
+        This method is overwritten (monkey-patched) in __enter__ for performance reasons
+        """
+        raise NotImplementedError
+
+    def seek(self, offset, whence=0):
+        return self.file.seek(offset, whence)
+
+    def tell(self):
+        return self.file.tell()
+
+    def __enter__(self):
+        self.file = open(self.path_, self.mode)
+        self.write = self.file.write
+        self.read = self.file.read
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.file.close()
+        if self.mod_time_to_set is not None:
+            set_file_mtime(self.path_, self.mod_time_to_set)
+
+
+class DownloadedFile:
+    def __init__(
+        self,
+        file_version: FileVersion,
+        strategy: AbstractDownloader,
+        range_: Optional[Tuple[int, int]],
+        response: Response,
+        encryption: Optional[EncryptionSetting],
+        session: B2Session,
+        progress_listener: AbstractProgressListener,
+    ):
+        self.file_version = file_version
+        self.strategy = strategy
+        self.range_ = range_
+        self.progress_listener = progress_listener
+        self.response = response
+        self.encryption = encryption
+        self.session = session
+
+    def _validate_download(self, bytes_read, actual_sha1):
+        if self.range_ is None:
+            if bytes_read != self.file_version.size:
+                raise TruncatedOutput(bytes_read, self.file_version.size)
+
+            if self.file_version.content_sha1 != 'none' and actual_sha1 != self.file_version.content_sha1:
+                raise ChecksumMismatch(
+                    checksum_type='sha1',
+                    expected=self.file_version.content_sha1,
+                    actual=actual_sha1,
+                )
+        else:
+            desired_length = self.range_[1] - self.range_[0] + 1
+            if bytes_read != desired_length:
+                raise TruncatedOutput(bytes_read, desired_length)
+
+    def save(self, file, mod_time_callback=None):
+        """
+        Read data from B2 cloud and write it to a file-like object
+        :param file: a file-like object
+        :param mod_time_callback: a callable accepting a single argument: the mod time of the downloaded file in milliseconds
+        """
+        if self.progress_listener:
+            file = WritingStreamWithProgress(file, self.progress_listener)
+            if self.range_ is not None:
+                total_bytes = self.range_[1] - self.range_[0] + 1
+            else:
+                total_bytes = self.file_version.size
+            self.progress_listener.set_total_bytes(total_bytes)
+        if mod_time_callback is not None:
```

Comment:
I was really confused what `mod_time_callback` actually does - it is very specific for `MtimeUpdatedFile` and I think moving this logic to `save_to`:
```
with MtimeUpdatedFile(path_, mode, mod_time=self.file_version.mod_time_millis) as file:
```
would be much better

like we give sdk user an object where he can access `file_version` (or `download_version` :P ) and `save` method - so he can handle `mod_time` by himself - really there is no need for this callback - especially when it can be confusing

Reply:
I think you're right. I don't really know why I did it this way (the over-engineered way:))